Fix loading variables from file with multi-system

DanielYankura opened the pull request:
-Fixes issue where initializing multiple variables from output file would fail 
-Nonlinear variables were not being properly copied to their corresponding system 
-A similar fix may be needed if the restart system is needed for multiple aux variables

closes #26580

Commit:
unrecognized variables when using output file to initialize variables

-Fixes issue where initializing multiple variables from output file would fail
-Nonlinear variables were not being properly copied to their corresponding system
-A similar fix may be needed if the restart system is needed for multiple aux variables

closes #26580

fixed spacing issues

applied style changes recommended by civit

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2410258/) on 68537b8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28475/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28475/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 307442af0570dad87c80f32022bb25380562cd28`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2410353/) on 069031a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28475/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2410368/) on 069031a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>307442</th>
      <th colspan=4>#28475 069031</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.02%</td>
      <td>85.02%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>105162</td>
      <td>105167</td>
      <td>+5</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18524</td>
      <td>18524</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28475/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28475/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28475/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -352,6 +352,12 @@ class FEProblemBase : public SubProblem, public Restartable
+  /**
+   * @return The number of nonlinear systems
+   */
+
```
```suggestion
```

GiudGiud commented on framework/src/actions/CopyNodalVarsAction.C:
```diff
@@ -57,11 +57,26 @@ CopyNodalVarsAction::act()
+        }
+      }
       else
+      {
         system = &_problem->getAuxiliarySystem();
+      }
```
let's not add those { }

GiudGiud commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -352,6 +352,12 @@ class FEProblemBase : public SubProblem, public Restartable
+   * @return The number of nonlinear systems
+   */
+
+  unsigned int getNonLinearSize() const { return _nl.size(); }
```
we have this api already?
```
  virtual std::size_t numNonlinearSystems() const override { return _num_nl_sys; }
```

DanielYankura replied:
I removed it from FEProblemBase.h and used the other function instead to get the number of nonlinear systems.

GiudGiud replied:
seems to be back now?

DanielYankura replied:
I think I made a mistake when pushing my changes because my branch didn't seem to have that function, so I recommitted and tried pushing again. Hopefully that fixes the issue.

GiudGiud commented on framework/src/actions/CopyNodalVarsAction.C:
```diff
@@ -57,11 +57,26 @@ CopyNodalVarsAction::act()
+      {
+        // This iterates through each nonlinear system and finds which one the current variable
+        // needs to be copied to
         system = &_problem->getNonlinearSystemBase(/*nl_sys=*/0);
```
no need to set this if you are going to loop after to find the real one

DanielYankura replied:
My thought with this was that it should be set to the first nonlinear system by default. That way if it is in fact a typo and can't be assigned to any of them, then when it tries to copy later it should throw an error instead of not trying to copy it. If it's better though I can have it throw an error right there it it isn't able to assign the variable to any of the nonlinear systems?

GiudGiud replied:
that works

DanielYankura replied:
I did some testing and it looks like assigning a variable by default to the first nonlinear system if unnecessary. If the user does have typos in their variable names will cause errors elsewhere so having a check there to make sure it was assigned doesn't do anything, so I made the changes you recommended.

GiudGiud reviewed (changes requested):
this looks like the right solution to this problem

we'll want to add a test for this

Commit:
Last of changes requested in PR review.
-Removed redundant function in FEProblemBase.h

Commit:
Small change to pass Civit test
-Added back in initialization of system variable on line 66
-Should help this pass Civet tests

Commit:
Added test

-starting_problem.i outputs an exodus file used as a restarting point for restart_problem.i
-restar_problem.i slightly changes the BCs and reruns
-outputs for both are tested

Updated restart test for multiple non-linear variables

-Removed initialization test as it was redundant (deletes starting_problem.i)
-Now restart_problem.i is restarted using gold/starting_problem_out.e and compared to restart_problem_out.e
-restart_problem.i no longer solves so only the copying/restarting is being tested
-tests file was updated to reflect these changes

Removed commented out line for restart_problem.i

Updates to restart test

-This test now uses the file used for restarting, as the comparison file used by exodiff
-restart_problem.i now includes execute_on=FINAL
-This ensures that the output file is just a copy of the input file

GiudGiud commented on test/tests/misc/multiple-nl-systems/restart_problem.i:
```diff
@@ -0,0 +1,77 @@
+[Problem]
+  nl_sys_names = 'u v'
+  allow_initial_conditions_with_restart = true
+[]
```
```suggestion
  solve = false
[]
```
then you can use the same exodus

GiudGiud replied:
(with a symlink)

DanielYankura replied:
I had a thought. Right now the only real purpose of starting_problem.i is to create an output file that can be used as a starting point for restart_problem.i. What if instead I have only one test that starts from an output file in the gold folder, does the copying but doesn't run, and then I can just have a check to make sure nothing changed during the restart process?

GiudGiud replied:
this would work I agree

GiudGiud commented on test/tests/misc/multiple-nl-systems/starting_problem.i:
```diff
@@ -0,0 +1,75 @@
+
+[Outputs]
+  exodus = true
+[]
```
```suggestion
  execute_on = FINAL
[]
```

DanielYankura replied:
This file was removed in the latest commit.

GiudGiud reviewed (commented):
looks good.

let s squash some commits together to make the "check in this file, then delete this file" disappear from the commit history (it s wasted storage)

DanielYankura commented:
I squashed together the three commits that added the test. The input file that was added, and then removed, shouldn't be in the commit history anymore.

GiudGiud commented on test/tests/misc/multiple-nl-systems/restart_problem.i:
```diff
@@ -0,0 +1,78 @@
+[Mesh]
+  file = 'gold/starting_problem_out.e'
```
cant you use restart_problem_out for the inital file?

DanielYankura replied:
The issue I found is that when restarting, even if no calculation is done, is that it will add a 2nd timestep that is the same as the first. So when you go to compare input and output files one has a time of 1 and the other 2 and exodiff fails. That's why I have it read in starting_problem_out.e, and compare its output to restart_problem_out.e.

DanielYankura replied:
I think start_time might only work for transient solvers? I tried using it but the test then errors because start_time is never used. I guess saying that it adds a 2nd timestep isn't accurate, but it adds a 2nd result. If there's a way to have a restart overwrite the previous output file instead of adding to it that might work. Otherwise the two output files might be necessary?

GiudGiud replied:
there is a `time` parameter for steady

it should not be appending, you are not outputting to gold/starting_problem.e

DanielYankura replied:
I tried messing around with the time parameter and I wasn't able to get it to quite work but I think I found a possible solution? I don't know if this is the right approach but adding execute_on = FINAL to the outputs block stops it from doing any execution after reading in the exodus file, so it should just read in the exodus file, copy over the solutions, and then output it right back out. Now when I compare its output against the file it reads in it matches. If that sounds like a reasonable solution then I can go ahead and push those changes.

GiudGiud replied:
That sounds good
I suspect execute_on = initial is all you need to compare the old exodus with the just read then output exodus

DanielYankura replied:
When I used execute_on = initial the output file was different from the input. I ended up using execute_on = final and pushed that change.

GiudGiud reviewed (changes requested):
the last but one commit still checks in two files. Let s squash them

you should also squash the meaningless format change commits into the one commit that introduced that code

we avoid squashing format commits only when they pertain to multiple commits (as it muddles the commit history instead of cleaning it)

GiudGiud commented on test/tests/misc/multiple-nl-systems/tests:
```diff
@@ -70,4 +70,14 @@
     requirement = 'The system shall be able to output postprocessor data after solving multiple solver systems.'
   []
+  [restart]
+    requirement = 'The system shall be able to restart from an exodus file containing multiple nonlinear variables and multiple nonlinear systems'
```
```suggestion
    requirement = 'The system shall be able to restart multiple nonlinear variables from different nonlinear systems  from an exodus file'
```
the exodus file does not have systems afaik (I dont know for sure but never seen it)

DanielYankura replied:
Just updated the wording!

GiudGiud reviewed (commented):
one tiny thing then good to go

Commit:
Changed test requirement wording for better accuracy

GiudGiud commented:
Awesome. Multi-system is the next big step, I m glad to see you support this
